Feature: Add LinearERC20VotingV1ValidatorV1

adamgall opened the pull request:
Depends on https://github.com/decentdao/decent-contracts/pull/254

Fixes [ENG-818](https://linear.app/decent-labs/issue/ENG-818/develop-validator-for-linearerc20votingv1)

**Summary:**

This PR introduces `LinearERC20VotingV1ValidatorV1`, a dedicated validator contract designed to work with `DecentPaymasterV1`. This validator specifically checks if a `vote(uint32, uint8)` UserOperation targeting a `LinearERC20VotingV1` strategy contract is likely to succeed before the paymaster agrees to sponsor it.

The PR includes the validator contract itself, a mock version of `LinearERC20VotingV1` for testing, and a comprehensive test suite for the new validator.

**Key Components:**

1.  **New Contract `LinearERC20VotingV1ValidatorV1.sol`:**
    *   Implements `IFunctionValidator`, `ERC165`, and `Version`.
    *   **`validateOperation(...)`:**
        *   Verifies the `callData` selector matches `ILinearERC20VotingV1.vote.selector`.
        *   Decodes `proposalId` and `voteType` from `callData`.
        *   Checks for valid `voteType` (0, 1, or 2).
        *   Retrieves proposal period (`startBlock`, `endBlock`) from the target `votingContract`.
        *   Ensures the proposal exists (non-zero `endBlock`).
        *   Verifies the voting period has not ended.
        *   Checks if the `lightAccountOwner` has already voted on the proposal.
        *   Retrieves the `governanceToken` from the `votingContract`.
        *   Fetches the `lightAccountOwner`'s checkpoints from the `governanceToken`.
        *   Returns `false` if the voter has no checkpoints.
        *   Iterates through checkpoints to find the voting weight at the `proposal.startBlock`.
        *   Returns `false` if the voting weight at `startBlock` is zero.
        *   Returns `true` if all checks pass, indicating the vote operation is likely to succeed.
    *   **`getVersion()`:** Returns the contract version (1).
    *   **`supportsInterface(...)`:** Implements ERC165 support for `IFunctionValidator`, `IVersion`, and `IERC165`.

2.  **New Mock Contract `MockLinearERC20VotingV1.sol`:**
    *   Provides a mock implementation of the functions from `LinearERC20VotingV1` and `IERC20Votes` that the validator interacts with.
    *   Allows setting proposal periods, voting ended status, voter status (`hasVoted`), governance token address, and checkpoint data for test scenarios.

3.  **New Test Suite `LinearERC20VotingV1ValidatorV1.test.ts`:**
    *   Provides comprehensive test coverage for `LinearERC20VotingV1ValidatorV1`.
    *   Tests various scenarios for `validateOperation`, including:
        *   Incorrect function selector in `callData`.
        *   Invalid vote types.
        *   Non-existent proposals.
        *   Voting period ended.
        *   User has already voted.
        *   Voter has zero voting weight (no checkpoints, all checkpoints after proposal start, relevant checkpoint has zero votes).
        *   Valid vote operations with various single and multiple checkpoint configurations.
        *   Edge cases like max voting weight and proposal start at block 0.
    *   Includes tests for ERC165 interface support and versioning.

**Motivation:**

*   To enable gasless voting for users interacting with `LinearERC20VotingV1` strategies via `DecentPaymasterV1`.
*   To provide a robust pre-validation mechanism that ensures the paymaster only sponsors vote operations that are likely to succeed, saving gas and preventing unnecessary sponsored transactions.
*   To create a reusable and upgradeable validator specific to the `LinearERC20VotingV1` contract's voting logic.

dacebt reviewed (commented):
I'm stuck on these validators. Where are these used deployed/setup/used in the Gasless voting setup

dacebt reviewed (approved):
Thanks for the InG update, makes sense. Curious just had a thought this is seems very specific to voting and you said that only one validator will be assigned. Will this pattern allow for gasless options in the future?

adamgall commented:
> Thanks for the InG update, makes sense. Curious just had a thought this is seems very specific to voting and you said that only one validator will be assigned. Will this pattern allow for gasless options in the future?

One validator is assigned _per target contract and function_. So yes, this pattern allows for additional transaction sponsorship (more than just voting) in the future.

johnqh commented on contracts/deployables/account-abstraction/validators/LinearERC20VotingV1ValidatorV1.sol:
```diff
@@ -0,0 +1,163 @@
+                lightAccountOwner,
+                uint32(i - 1)
+            );
+
```
I was thinking about this and there is another optimization we can do:
if (checkpoint.fromBlock > endBlock) {
// This means the vote has ended, since there is a checkpoint after the proposal endBlock. We can break out of the loop and return false.
}

adamgall replied:
updated

johnqh reviewed (changes requested):
Additional optimization we can do

adamgall committed:
Refactor: Simplify LinearERC20VotingV1ValidatorV1 Contract

**Summary:**

This commit refactors the `LinearERC20VotingV1ValidatorV1` contract by removing the UUPS and Ownable upgradeability features, streamlining its implementation. The `initialize` function and upgrade authorization logic have been eliminated, simplifying the contract's structure while maintaining its core validation functionality.

**Key Changes:**

1. **Contract Simplification:**
   - Removed inheritance from `UUPSUpgradeable` and `OwnableUpgradeable`.
   - Deleted the `initialize` function and the `_authorizeUpgrade` method.

2. **Test Adjustments:**
   - Updated the test suite to reflect the removal of the initialization and upgradeability logic, ensuring continued coverage of the validator's functionality.

**Motivation:**

* To enhance the clarity and maintainability of the `LinearERC20VotingV1ValidatorV1` contract by reducing complexity.
* To focus on the core validation logic without the overhead of upgradeability features that are not currently needed.

adamgall committed:
Feature: Add LinearERC20VotingV1ValidatorV1

**Summary:**

This PR introduces `LinearERC20VotingV1ValidatorV1`, a dedicated validator contract designed to work with `DecentPaymasterV1`. This validator specifically checks if a `vote(uint32, uint8)` UserOperation targeting a `LinearERC20VotingV1` strategy contract is likely to succeed before the paymaster agrees to sponsor it.

The PR includes the validator contract itself, a mock version of `LinearERC20VotingV1` for testing, and a comprehensive test suite for the new validator.

**Key Components:**

1.  **New Contract `LinearERC20VotingV1ValidatorV1.sol`:**
    *   Implements `IFunctionValidator`, `ERC165`, `Version`, `UUPSUpgradeable`, and `OwnableUpgradeable`.
    *   **`initialize(address owner)`:** Sets up ownership.
    *   **`validateOperation(...)`:**
        *   Verifies the `callData` selector matches `ILinearERC20VotingV1.vote.selector`.
        *   Decodes `proposalId` and `voteType` from `callData`.
        *   Checks for valid `voteType` (0, 1, or 2).
        *   Retrieves proposal period (`startBlock`, `endBlock`) from the target `votingContract`.
        *   Ensures the proposal exists (non-zero `endBlock`).
        *   Verifies the voting period has not ended.
        *   Checks if the `lightAccountOwner` has already voted on the proposal.
        *   Retrieves the `governanceToken` from the `votingContract`.
        *   Fetches the `lightAccountOwner`'s checkpoints from the `governanceToken`.
        *   Returns `false` if the voter has no checkpoints.
        *   Iterates through checkpoints to find the voting weight at the `proposal.startBlock`.
        *   Returns `false` if the voting weight at `startBlock` is zero.
        *   Returns `true` if all checks pass, indicating the vote operation is likely to succeed.
    *   **`getVersion()`:** Returns the contract version (1).
    *   **`supportsInterface(...)`:** Implements ERC165 support for `IFunctionValidator`, `IVersion`, and `IERC165`.

2.  **New Mock Contract `MockLinearERC20VotingV1.sol`:**
    *   Provides a mock implementation of the functions from `LinearERC20VotingV1` and `IERC20Votes` that the validator interacts with.
    *   Allows setting proposal periods, voting ended status, voter status (`hasVoted`), governance token address, and checkpoint data for test scenarios.

3.  **New Test Suite `LinearERC20VotingV1ValidatorV1.test.ts`:**
    *   Provides comprehensive test coverage for `LinearERC20VotingV1ValidatorV1`.
    *   Tests various scenarios for `validateOperation`, including:
        *   Incorrect function selector in `callData`.
        *   Invalid vote types.
        *   Non-existent proposals.
        *   Voting period ended.
        *   User has already voted.
        *   Voter has zero voting weight (no checkpoints, all checkpoints after proposal start, relevant checkpoint has zero votes).
        *   Valid vote operations with various single and multiple checkpoint configurations.
        *   Edge cases like max voting weight and proposal start at block 0.
    *   Includes tests for ERC165 interface support, versioning, and UUPS upgradeability.

**Motivation:**

*   To enable gasless voting for users interacting with `LinearERC20VotingV1` strategies via `DecentPaymasterV1`.
*   To provide a robust pre-validation mechanism that ensures the paymaster only sponsors vote operations that are likely to succeed, saving gas and preventing unnecessary sponsored transactions.
*   To create a reusable and upgradeable validator specific to the `LinearERC20VotingV1` contract's voting logic.

adamgall committed:
Enhancement: Add Voting Period End Check to LinearERC20VotingV1ValidatorV1

**Summary:**

This commit enhances the `LinearERC20VotingV1ValidatorV1` contract by introducing a check to invalidate votes if the most recent checkpoint occurs after the proposal's `endBlock`. This ensures that any voting attempts made after the proposal has ended are correctly rejected.

**Key Changes:**

1. **Contract Logic Update:**
   - Added a condition in the `validateOperation` function to return `false` if the latest checkpoint's block is greater than the proposal's `endBlock`, indicating that the voting period has concluded.

2. **Test Suite Expansion:**
   - Introduced new test cases to verify the behavior of the validator when checkpoints are set after the proposal's end, ensuring that votes are invalidated as expected.

**Motivation:**

* To improve the integrity of the voting process by ensuring that votes cannot be cast after the designated voting period has ended, thereby aligning with governance best practices.

adamgall committed:
Refactor: Update LinearERC20VotingV1 and Validator to Use Timestamps

**Summary:**

This commit refactors the `LinearERC20VotingV1` and `LinearERC20VotingV1ValidatorV1` contracts to replace block-based logic with timestamp-based logic for proposal periods and voting checkpoints. This change enhances the accuracy of voting period management and aligns with modern practices in smart contract development.

**Key Changes:**

1. **Contract Modifications:**
   - Updated `getProposalPeriod` to return timestamps instead of block numbers.
   - Refactored `Checkpoint` structures to use `uint48` for timestamps and `uint208` for voting weights.
   - Adjusted all relevant functions to utilize timestamps for proposal validation and voting weight calculations.

2. **Mock Contract Updates:**
   - Modified `MockERC20Votes` and `MockLinearERC20VotingV1` to support timestamp-based voting logic.

3. **Test Suite Enhancements:**
   - Updated tests in `LinearERC20VotingV1ValidatorV1.test.ts` to reflect the changes from block-based to timestamp-based logic, ensuring comprehensive coverage of the new functionality.

**Motivation:**

* To improve the precision of voting periods and enhance the overall integrity of the voting process by utilizing timestamps, which are more reliable in the context of smart contracts.
* To ensure compatibility with future developments in the Ethereum ecosystem that favor timestamp-based operations over block-based ones.
